File: sdk/typescript/examples/saga-pattern.ts
/**
 * Example: Multi-step saga with compensation using the VELOCITY-WorkFlow TypeScript SDK.
 *
 * Demonstrates:
 *   - Defining a saga with compensable steps
 *   - Executing steps in order
 *   - Triggering compensation on failure
 *   - Rolling back completed steps in reverse order
 *
 * Prerequisites:
 *   1. Start the VELOCITY-WorkFlow server:
 *      cd VELOCITY-WorkFlow/src/Velocity.Workflow.Server && dotnet run
 *   2. npm install
 *   3. npx ts-node examples/saga-pattern.ts
 */

import { VelocityClient, WorkflowStub, WorkflowStatus } from '../src';

// ── Saga step definitions ────────────────────────────────────────────────────

interface SagaStep {
  name: string;
  compensate: string;
}

const STEPS: SagaStep[] = [
  { name: 'reserve_inventory', compensate: 'release_inventory' },
  { name: 'charge_payment',    compensate: 'refund_payment' },
  { name: 'book_shipping',     compensate: 'cancel_shipping' },
  { name: 'send_confirmation', compensate: 'send_cancellation_notice' },
];

async function executeStep(
  stub: WorkflowStub,
  step: SagaStep,
): Promise<boolean> {
  console.log(`   Executing: ${step.name}`);
  await stub.signal(step.name, {});
  return true;
}

async function compensateStep(stub: WorkflowStub, step: SagaStep): Promise<void> {
  console.log(`   Compensating: ${step.compensate}`);
  await stub.signal(step.compensate, {});
}

async function runSaga(
  client: VelocityClient,
  simulateFailureAt?: number,
): Promise<boolean> {
  const stub = new WorkflowStub(client, {
    workflowType: 'order-saga',
    namespace: 'default',
    taskQueue: 'orders',
  });

  const handle = await stub.start({ sagaId: Date.now() });
  console.log(`  Saga started: key=${handle.workflowKey}`);

  const completedSteps: SagaStep[] = [];

  for (let i = 0; i < STEPS.length; i++) {
    if (simulateFailureAt !== undefined && i === simulateFailureAt) {
      console.log(`\n   ✗ Step '${STEPS[i].name}' FAILED — triggering compensation`);
      for (const prev of [...completedSteps].reverse()) {
        await compensateStep(stub, prev);
      }
      await stub.terminate(`Step ${STEPS[i].name} failed`);
      return false;
    }

    const success = await executeStep(stub, STEPS[i]);
    if (success) completedSteps.push(STEPS[i]);
  }

  console.log('   ✓ All saga steps completed successfully');
  return true;
}

async function main(): Promise<void> {
  console.log('=== VELOCITY-WorkFlow TypeScript SDK — Saga Pattern ===\n');

  const client = new VelocityClient('localhost:50051');
  await client.connect();

  // Scenario 1: Happy path
  console.log('Scenario 1: Happy path');
  await runSaga(client);

  // Scenario 2: Payment step fails (index=1)
  console.log('\nScenario 2: Payment step fails (index=1)');
  await runSaga(client, 1);

  await client.close();
  console.log('\n=== Saga examples finished! ===');
}

main().catch(console.error);

File: sdk/rust/examples/saga_pattern.rs
//! Example: Multi-step saga with compensation using the VELOCITY-WorkFlow Rust SDK.
//!
//! Demonstrates:
//!   - Defining a saga with compensable steps
//!   - Executing steps in order
//!   - Triggering compensation on failure
//!   - Rolling back completed steps in reverse order
//!
//! Run:
//!   cd VELOCITY-WorkFlow/sdk/rust
//!   cargo run --example saga_pattern

use velocity_sdk::{VelocityClient, WorkflowStatus};

/// A saga step with a forward action and a compensation action.
struct SagaStep {
    name: &'static str,
    compensate: &'static str,
    signal_id: u64,
    compensate_id: u64,
}

const STEPS: &[SagaStep] = &[
    SagaStep { name: "reserve_inventory", compensate: "release_inventory", signal_id: 100, compensate_id: 200 },
    SagaStep { name: "charge_payment",    compensate: "refund_payment",    signal_id: 101, compensate_id: 201 },
    SagaStep { name: "book_shipping",     compensate: "cancel_shipping",   signal_id: 102, compensate_id: 202 },
    SagaStep { name: "send_confirmation", compensate: "send_cancellation", signal_id: 103, compensate_id: 203 },
];

/// Run the saga. If `simulate_failure_at` is Some(i), step i will fail.
fn run_saga(client: &VelocityClient, simulate_failure_at: Option<usize>) -> bool {
    let key = client.start_workflow(1, 1, 42, STEPS.len() as u32);
    assert!(key > 0);
    println!("  Saga started: key={key}");

    let mut completed: Vec<usize> = Vec::new();

    for (i, step) in STEPS.iter().enumerate() {
        // Simulate failure
        if simulate_failure_at == Some(i) {
            println!("\n   ✗ Step '{}' FAILED — triggering compensation", step.name);
            // Compensate in reverse order
            for &prev_idx in completed.iter().rev() {
                let prev = &STEPS[prev_idx];
                println!("   Compensating: {}", prev.compensate);
                client.signal_workflow(key, prev.compensate_id, prev.compensate.as_bytes().to_vec());
            }
            return false;
        }

        println!("   Executing: {}", step.name);
        client.signal_workflow(key, step.signal_id, step.name.as_bytes().to_vec());
        completed.push(i);
    }

    println!("   ✓ All saga steps completed successfully");
    true
}

fn main() {
    println!("=== VELOCITY-WorkFlow Rust SDK — Saga Pattern ===\n");

    let client = VelocityClient::new();

    // Scenario 1: Happy path
    println!("Scenario 1: Happy path");
    run_saga(&client, None);

    // Scenario 2: Payment step fails (index=1)
    println!("\nScenario 2: Payment step fails (index=1)");
    run_saga(&client, Some(1));

    client.destroy();
    println!("\n=== Saga examples finished! ===");
}
